doc/developer/design/20231211_datadog_sink.md

```diff
@@ -0,0 +1,162 @@
+    - https://github.com/MaterializeInc/materialize/issues/17601
+    - https://github.com/MaterializeInc/materialize/issues/4779
+
+## The Problem
+
+Users want to export **metrics** to Datadog. But they need to learn how to configure, test, deploy, and maintain two external services: the Datadog Agent and a SQL Exporter — making the experience long and sometimes even frustrating.
+
+## Success Criteria
+
+In a few clicks, a user can integrate Materialize with Datadog. Also, the user can export custom metrics to Datadog by running the following SQL commands:
+
+```sql
+CREATE CONNECTION conn_datadog TO DATADOG API KEY = ...;
+CREATE SINK FROM <view/table> TO DATADOG CONNECTION conn_datadog;
+```
+
+The source content of the sink, which can be either a view or a table, will be defined by the user but needs to match a predefined structure.
+
+## Out of Scope
+
+It is out of the scope:
+
+- Alternative approaches, such as https://github.com/MaterializeInc/materialize/issues/4779, to select labels and values, or send any additional information to Datadog outside the connection validation or metrics.
+- Compression algorithms before exporting metrics.
+
+## Solution Proposal
+
+The solution proposal involves a new connection and sink for Datadog, and an OAuth process to create a seamless experience.
+
+### Datadog Connection
+
+The new connection will store Datadog’s API key, which is necessary to submit metrics to Datadog. The first time the user runs the integration, an API key will be created automatically by the console (refer to the [OAuth process diagram](#oauth-process) for more details.)
+
+Command:
+```sql
+CREATE CONNECTION conn_datadog TO DATADOG API KEY = ...;
+```
+
+### Datadog Sink
+
+We have two possible approaches, with the first one being simpler:
+
+1. The new sink will query the source, a table or view, at a 15-second interval and submit the results as metrics to Datadog. In case the sink crashes or gets paused, the sink will send the current metric values in the table or view.
+2. Build a state inside the sink, and submit the whole state to Datadog at a 15-second interval. This approach benefits from less query overhead but requires memory to keep/duplicate the state.
+
+For both approaches, the 15-second interval comes from [the Datadog agent](https://docs.datadoghq.com/agent/basic_agent_usage/?tab=agentv6v7#collector) collection interval. This value could be customizable but as a first approach and MVP, a default value is ok. If the sink crashes or pauses, it will resume from the newest metric values, skipping any missing values. We will use the wall-clock time, so it wouldn't be justified to rewind unless we want to guarantee the correctness and use Materialize internal clock or go back in history to fill in the missing points. I would qualify this feature as a nice-to-have but not a minimum requirement. This mechanism is the same as the current integrations and how the Grafana integration will also work.
```

Comment:
I think it will actually be much easier to structure the code so that you send metrics for every 15s interval to Datadog, and you'll get catchup "for free" based on how sinks are already caught back up when Materialize starts.

Reply:
Yeah. After more thinking, sticking to the same logic as the Kafka sink brings way more coherence to the code, making it also simpler to maintain in the long run, etc. From what I saw in the code, this approach will also inherit compaction hold in case of failure and resume when the frontier recovers.